src/cdogs/keyboard.c

```diff
@@ -107,6 +108,93 @@ void KeyOnKeyUp(keyboard_t *keyboard, const SDL_Keysym s)
 	keyboard->currentKeys[s.scancode].isPressed = 0;
 }
 
+void DiagonalHold(keyboard_t *keyboard, int c)
+{
+    //To change the level of delay, set DIAGONAL_RELEASE_DELAY up top to another number, if more than two players are ever allowed to map to a keyboard, expansion should just be changing the for loop in KeyPostPoll that increments c
+    // ------
+    int currentTicks = (int)SDL_GetTicks(); //Used to determine whether to keep holding a diagonal by comparing how long it has been since it set that diagonal's ticks vs current time
```

Comment:
don't use `SDL_GetTicks` directly; this function should take `ticks` as an argument, which is passed in from the game loop, where `ticks` is the elapsed number of game ticks (usually just 1, could be 2 under frame skipping).

So instead of getting the current ticks and checking if the diagonal ticks is greater than that, you would initialise diagonal ticks to 0, set it to `DIAGONAL_RELEASE_DELAY` when the diagonal is first pressed, and decrement it by `ticks` every time this function is called, and when it's <= 0 the sustain ends. This pattern is used in lots of other places, e.g. see `soundLock`.

Reply:
by the way, one `tick` is 1/70th of a second, defined by the game tick rate as `FPS_FRAMELIMIT`. So if you're looking for a conversion between 20ms into ticks, it's about 1 and a half ticks.

Reply:
Actually now that I think about it, this delay should be independent of the game tick rate. Don't worry about this then; using `SDL_GetTicks` is fine.